proposals/simd/SIMD.md

```diff
@@ -285,14 +285,40 @@ The input lane value, `x`, is interpreted the same way as for the splat
 instructions. For the `i8` and `i16` lanes, the high bits of `x` are ignored.
 
 ### Shuffle lanes
+
+#### Immediate permutation rule
 * `v8x16.shuffle(a: v128, b: v128, imm: ImmLaneIdx32[16]) -> v128`
 
 Returns a new vector with lanes selected from the lanes of the two input vectors
-`a` and `b` specified in the 16 byte wide immediate mode operand `imm`. This
-instruction is encoded with 16 bytes providing the indices of the elements to
+`a` and `b` specified in the 10 byte wide immediate mode operand `imm`. This
+instruction is encoded with 10 bytes providing the indices of the elements to
```

Comment:
I assumed the immediate operand would be packed.
I haven't verified such assumptions, and can revert back to 1 byte per index for this PR (and consider the packing in another PR).

My rationale about this is how wasteful the non-packed is.
In the following table, you can see how many bytes are used by the immediate in a packed and unpacked scheme:

| instruction | unpacked | packed | diff |
|:-------------:|:------------:|:---------:|:-----:|
| `v8x16.shuffle` | 16 | 10 | 6 |
| `v16x8.shuffle` | 8 | 3 | 5 |
| `v32x4.shuffle` | 4 | 2 | 2 |
| `v64x2.shuffle` | 2 | 1 | 1 |
| `v8x16.permute`* | 16 | 8 | 8 |
| `v16x8.permute`* | 8 | 2 | 6 |
| `v32x4.permute`* | 4 | 1 | 3 |
| `v64x2.permute`* | 2 | 1 | 1 |

*: not integrated in this PR

As you can see, the difference is significant for `v8x16`, where the packed version is 6 bytes shorter.

I also claim it is not really difficult to implement in the virtual machine:
```c
struct packed5x8 {
    char i0:5;
    char i1:5;
    char i2:5;
    char i3:5;
    char i4:5;
    char i5:5;
    char i6:5;
    char i7:5;
};

void unpack5x8(const char* packed, char* unpacked) {
    struct packed5x8 const* unpacker = (struct packed5x8 const*) packed;
    unpacked[0x0] = unpacker->i0;
    unpacked[0x1] = unpacker->i1;
    unpacked[0x2] = unpacker->i2;
    unpacked[0x3] = unpacker->i3;
    unpacked[0x4] = unpacker->i4;
    unpacked[0x5] = unpacker->i5;
    unpacked[0x6] = unpacker->i6;
    unpacked[0x7] = unpacker->i7;
}
void unpack5x16(const char* packed, char* unpacked) {
    unpack5x8(packed, unpacked);
    unpack5x8(packed+5, unpacked+8);
}
```

Reply:
> My rationale about this is how wasteful the non-packed is.

That's the main pro I could find.

> I also claim it is not really difficult to implement in the virtual machine:

The complexity appears in the generated core, where packing / unpacking is more expensive that just doing a memcpy.

Reply:
> The complexity appears in the generated core, where packing / unpacking is more expensive that just doing a memcpy.

I know the generated code is more expensive, but I tend to think it is negligible compared to the code transmission or the other steps of the compilation.

IMO, the size reduction outweights the translation time cost.

Also, I'm fine putting this in another PR.

Reply:
I think it would be better to open up a discussion issue about this, mentioning all the pros (less code size, minimal implementation cost) and cons (increased cost of serializing / deserializing the binary format), so that people can form an informed opinion.

Reply:
I'm skeptical of adding more cost to generated code for shuffles, as the generic shuffle is already going to be slow on some platforms. Add this to the fact that  there is a high likelihood that any operations that are missing from this portable subset will need to be emulated using shuffles, this disproportionately increases the cost of the generated code.
I agree with gnzlbg@, that this needs it's own issue. If there are other arguments to packing, aside from a general sense of wastefulness, please open a separate issue/PR. IMO, this is a trade off that we can't make without actual numbers, i.e. the cost of generated code being negligible is not a given.